src/hamiltonian.jl

```diff
@@ -1,8 +1,29 @@
 # TODO: add a type for kinetic energy
 
+# The constructor of `LogDensityValue` will check numerical errors in
+#   `value` and `gradient`.  That is `is_valid` will be performed automatically.
+struct LogDensityValue{Tv<:AbstractFloat, Tg<:AbstractVector{Tv}}
```

Comment:
Not sure if `LogDensityValue` is a good name:
- We are working on log-joint probability instead of simply `LogDensity`. But I'm OK with it since it's also a log density.
- We are storing both value and gradient here. Why do we only name it with `Value`?

Reply:
yeah, perhaps we could use the following 

- `DualValue`: store value and its corresponding gradients
- `DualFunction`: store function and its corresponding gradient function

This echoes the naming scheme of AD packages, so should be self-explanatory.